Initial contribution of the bespoke scheduler infrastructure for ESH V2.0

kgoderis opened the pull request:
Signed-off-by: Karel Goderis <karel.goderis@me.com>

kgoderis commented:
Stefan, Kai, this should be ok for a first review round.

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,240 @@
+ */
+public abstract class AbstractExpression<E extends AbstractExpressionPart> implements Expression {
+
+    final Logger logger = LoggerFactory.getLogger(this.getClass().getName());
```
this should be private

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,240 @@
+public abstract class AbstractExpression<E extends AbstractExpressionPart> implements Expression {
+
+    final Logger logger = LoggerFactory.getLogger(this.getClass().getName());
+
```
most of the below fields (if not all - see somments below...) can and should be private. I'd use protected getters instead - they make it easier to make it thread-safe

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,240 @@
+        applyExpressionParts();
+
+        continueSearch = true;
+        while (candidates.size() < minimumCandidates && continueSearch) {
```
Your are using a protected field for ```continueSearch``` here. As it's an instance variable and there are no synchronized blocks/statements, this will break when getting executed from different threads

kgoderis replied:
Yeah you are right. I have not written it with multithreading in mind. I will review the whole lot in that sense

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,240 @@
+
+    abstract protected void validateExpression() throws IllegalArgumentException;
+
+    @SuppressWarnings("unchecked")
```
not needed

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java:
```diff
@@ -0,0 +1,171 @@
+
+    @Override
+    public Date getFinalFireTime() {
+        if (candidates.isEmpty()) {
```
This block is repeated a couple of times - I think it would be good to factor it out into the abstract base class

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpressionPart.java:
```diff
@@ -0,0 +1,173 @@
+abstract class AbstractExpressionPart implements ExpressionPart {
+
+    protected String part;
+    protected BoundedIntegerSet valueSet;
```
again, I'd rather use the existing getter, making the field private

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,240 @@
+     * @throws ParseException - when the expression can not be successfully be parsed
+     * @throws IllegalArgumentException - when expression parts conflict with each other
+     */
+    @SuppressWarnings("unchecked")
```
not needed

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpressionPart.java:
```diff
@@ -0,0 +1,173 @@
+        this.valueSet = set;
+    }
+
+    abstract BoundedIntegerSet initialiseValueSet();
```
afaik we usually tend to use the american spelling here

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpressionPart.java:
```diff
@@ -0,0 +1,173 @@
+         * @param integer - the integer to validate
+         */
+        private void validate(final Integer integer) {
+            if (!negative) {
```
could you turn this around? ``ìf (negative) {...}``` is easier to read than ``ìf (!negative) {...}```

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/CronExpression.java:
```diff
@@ -0,0 +1,1369 @@
+ * @author Karel Goderis - Initial contribution
+ *
+ */
+public final class CronExpression extends AbstractExpression<CronExpressionPart> {
```
I'd prefer to use a dedicated package with several classes instead of one huge class with tons of inner classes. Maybe it's personal preference, but such a long class is harder to read and to maintain.

kgoderis replied:
Maybe.... I did it that way to keep the classes private as they are only relevant for - in this case - CronExpression.

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java:
```diff
@@ -0,0 +1,171 @@
+            // ISO8601 date string according to the XML Schema specification.
+
+            try {
+                Calendar cal = DatatypeConverter.parseDateTime(expression);
```
Just as a side-mark: The ```javax.xml.bind``` package is not available in the Compact 2 profile, which will make it yet harder to run any ESH based products on really weak hardware. So if there is another feasible way, I'd suggest to not draw in a dependency to jaxb here. Although I clearly understand how handy it is here...!

kgoderis replied:
Ok - let me find an alternative then. When I googled for a solution, there were little to no good ones, with all of them missing some part of the ISO8601 definitions. Of course, if we can live with this (and not support some of variations in ISO860), the alternative solution will be quickly found

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,240 @@
+     * @param startDate - the start date of the expression
+     * @param timeZone - the time zone of the expression
+     * @param minimumCandidates - the minimum number of candidate dates to calculate
+     * @throws ParseException - when the expression can not be parsed correctly
```
"can not" -> "cannot"

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,240 @@
+    @Override
+    public void setStartDate(Date startDate) {
+        if (startDate == null) {
+            throw new IllegalArgumentException("The start date of the rule can not be null");
```
"can not" -> "must not"

sjka commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,240 @@
+     * Parse the given expression
+     * 
+     * @param expression - the expression to parse
+     * @throws ParseException - when the expression can not be successfully be parsed
```
"can not" -> "cannot"

sjka commented:
From a legal point of view, that should be fine now as far as I can see.

There are a few minor remarks that I left in the code, I hope they make sense to you?

Generally I about the naming, the class names clearly make sense, but e.g. "Expression" is quite generic and it's currently located in the equally generic .core.common package - therefore _some_ namespacing would be good, e.g. by using a dedicated sub-package.

Functionality wise - it's hard to tell in theory. The logic you have implemented is a perfect target for unit-tests. And I don't feel comfortable with having such logic without any tests at all.  Would you happen to have some tests for your implementation too? I guess you must have somehow tried it out as well...

Regarding the PR, I think it would make sense to squash the commits in order to not have the first implementation with the IP issues merged.

kgoderis commented:
@skja I have never written groovy stuff in my life, so I am totally lost with respect to the unit testing. However, that being said, i have tested a ton of expressions for both cron and rfc5545 "manually". These were expressions I copied from websites that also contained the supposed output, and I compared to that to validate the correctness of the code. 

Expression we could change to SchedulerExpression or SchedulingExpression?

kgoderis commented:
@sjka, @kaikreuzer any next steps? This PR is necessary to be able to complete the Automation PR, which in turn is necessary for the CalDav stuff that is nearly finished

kaikreuzer commented:
> some namespacing would be good, e.g. by using a dedicated sub-package.

I agree - actually, we currently even have a separate bundle (.core.scheduler); if we argue that it is a central part that everybody will need and we want to reduce the number of bundles, I am also fine with putting the code in the .core bundle - but then it should be in the package .core.scheduler and not .core.common.

> I have never written groovy stuff in my life, so I am totally lost with respect to the unit testing. 

Note that tests do not have to be written in Groovy - pure Java is fine as well.
As there is already a .core.tests fragment, you would not even have to deal with the whole setup and skeleton - all there is to do is to add a few tests. Have a look at https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/library/types/PercentTypeTest.java as an example - this is pretty easy and straight forward. Maybe you want to have a try after all? :-)

@maggu2810 As this is a major code contribution, I would love to have your input and comments on this PR as well!

kgoderis commented:
> I have never written groovy stuff in my life, so I am totally lost with respect to the unit testing.
> Note that tests do not have to be written in Groovy - pure Java is fine as well.
> As there is already a .core.tests fragment, you would not even have to deal with the whole setup and skeleton - all there is to do is to add a few tests. Have a look at https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/library/types/PercentTypeTest.java as an example - this is pretty easy and straight forward. Maybe you want to have a try after all? :-)

How do I see or test that the @Test is working? Can I do that from within the Eclipse IDE?

K

kaikreuzer commented:
Yes, simply right-click the test class and choose "Run As -> JUnit Test".

kgoderis commented:
> some namespacing would be good, e.g. by using a dedicated sub-package.
> I agree - actually, we currently even have a separate bundle (.core.scheduler); if we argue that it is a central part that everybody will need and we want to reduce the number of bundles, I am also fine with putting the code in the .core bundle - but then it should be in the package .core.scheduler and not .core.common.

as it is, .core.scheduler could/should disappear no? as it only links in the Quartz libs.

FYI, I use it throughout the other stuff I mentioned (migration of DSL Script/Rule to Automation) as well as CalDav stuff, so I would expect it to be .core. e.g. a renewed .core.scheduler then

kaikreuzer commented:
Yes, I am fine with that. So .core bundle and .core.scheduler package it is.

maggu2810 commented:
> @maggu2810 As this is a major code contribution, I would love to have your input and comments on this PR as well!

Sure, it is already on my Todo list.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/CronExpression.java:
```diff
@@ -0,0 +1,1350 @@
+ */
+public final class CronExpression extends AbstractExpression<CronExpressionPart> {
+
+    private final static Logger logger = LoggerFactory.getLogger(CronExpression.class);
```
We are using non-static logger instances.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java:
```diff
@@ -0,0 +1,177 @@
+ */
+public class DateExpression extends AbstractExpression<DateExpressionPart> {
+
+    private final static Logger logger = LoggerFactory.getLogger(DateExpression.class);
```
We are using non-static logger instances.

maggu2810 replied:
As already stated, you should use a non-static logger instance if possible. In this case, we do not need the logger instance at all. Nothing is logged in the current code base.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/CoreActivator.java:
```diff
@@ -28,6 +29,10 @@
         context = bc;
         logger.debug("Core bundle has been started.");
+
+        DateExpression de = new DateExpression("2016-01-31T16:45:46+00:00");
```
What the intention of this line?

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/CoreActivator.java:
```diff
@@ -28,6 +29,10 @@
+
+        DateExpression de = new DateExpression("2016-01-31T16:45:46+00:00");
+
+        logger.debug("stop");
```
What the intention of this line?

kgoderis replied:
Wrongly committed file. test left over

maggu2810 commented:
This PR is a big one and to review it I want to use the code base on my machine (IDE etc.).
I checked this PR out and have done a <code>mvn clean install</code>.
This does not succeed caused by test failure: https://gist.github.com/maggu2810/6c4ca383d41326b38589

kgoderis commented:
@magg2810 need to check that, as it works at my side (in the IDE). That being said, could maybe due to timezone differences, and in that case something else needs fixing.

kgoderis commented:
> This PR is a big one and to review it I want to use the code base on my machine (IDE etc.).
> I checked this PR out and have done a mvn clean install.
> This does not succeed caused by test failure: https://gist.github.com/maggu2810/6c4ca383d41326b38589

Ok. fixed this test. my mistake, the RRULE test only works in US presidential election years. 2016 is, 2017 is not.

kaikreuzer commented:
What is the status of this PR? I see that the classes are still in core.common and not in core.scheduler, so I assume you are still working on it?
I also noticed that you will have to take out the expression related code from the ThreadPoolManager.
I'd suggest that I actually first port the new ThreadPoolManager implementation from jUPnP over and we can then see what you might need to extend for the expression support - my hope would be that you actually do not have to touch this class itself at all.

maggu2810 commented:
After the tests has been fixed, I have planed to start a review this weekend.
If the classes are not moved yet, I assume not all of comments of @kaikreuzer have been integrated.
If there is no need to merge this PR as fast as possible, I would wait for my review, after the already commented stuff has been fixed / changed.
Ok?

kgoderis commented:
I have not moved the classes yet, planned to do that another round of comments. All the rest is done except the last comment regarding the jupnp port

maggu2810 commented:
@kgoderis Okay, I will check out the PR locally and try to spend the next hour to look at your changes. Let's see how far I get.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/META-INF/MANIFEST.MF:
```diff
@@ -53,7 +53,6 @@ Import-Package: com.google.common.base,
  org.osgi.service.event,
  org.osgi.service.log,
  org.osgi.util.tracker,
- javax.xml.bind,
```
The package javax.xml.bind is used by <code>bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/RawType.java</code>
There is this import statement: <code>import javax.xml.bind.DatatypeConverter;</code>

kgoderis replied:
Ok - not sure why it was in the commit

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,278 @@
+ */
+public abstract class AbstractExpression<E extends AbstractExpressionPart> implements Expression {
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass().getName());
```
org.slf4j.LoggerFactory should support a class atrribute, is there any reason to use getName()?

kgoderis replied:
Nope

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,278 @@
+        this.timeZone = timeZone;
+        this.minimumCandidates = minimumCandidates;
+
+        setStartDate(startDate);
```
A constructor should not call non-final functions:
https://www.securecoding.cert.org/confluence/display/java/MET05-J.+Ensure+that+constructors+do+not+call+overridable+methods

maggu2810 replied:
Same for the other ones...

kgoderis replied:
Ok - that could be changed as there are no specific subclass implementations

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,278 @@
+
+    @Override
+    public final Date getStartDate() {
+        if (startDate == null) {
```
All fine, but I would assume that the getStartDate is called more often (e.g. more often than setStartDate).
If we ensure that startDate is never null (in the constructor and setStartDate function) we do not need this check every time getStartDate is called.

Or do you like that the time of the first getStartDate call is used?

maggu2810 replied:
Same for e.g. timezone

kgoderis replied:
I would leave it as it is, as not for every kind of Expression you need a startdate, and some kinds can work without a startDate.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/Expression.java:
```diff
@@ -0,0 +1,97 @@
+ * @author Karel Goderis
+ *
+ */
+public interface Expression {
```
Is there any need to add the setters?
I would assume that an expression is created and used, is there any use case you change e.g. expressen, time zone, start date etc. of an already existing expression object?

kgoderis replied:
Yes! The start date is the date from which you want the expresion to be evaluated. The start date is not perse encoded in that expression, so it should/could be set afterwards. The same is true for the timezone, as some expressions do not capture timezone information.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ExpressionPart.java:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @param s
+     */
+    public void set(String s);
```
Same here as above, would you like to reuse an ExpressionPart object and call set, parse multiple times or should the String be given by the constructor?

kgoderis replied:
Here it could be removed I think. Unless, for another kind of Expression, specific setters and getters are defined to change part of the definition. In that case, the existing ExpressionPart should be re-used. Since the ExpressionParts are created and added to an ArrayList, I prefer them to be changed instead of looking for the correct ExpressionPart in the list, remove it, create a new one and add that one instead...

maggu2810 commented:
FYI: I will add comments while reading, if there are some obsolete after reading other files, I will remove them later...

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadPoolManager.java:
```diff
@@ -175,6 +186,38 @@ static public ExecutorService getPool(String poolName) {
 
+    /**
+     * Returns an instance of an expression-driven scheduled thread pool service. If it is the first request for the
+     * given pool name, the
```
Perhaps you could remove this line break...

kgoderis replied:
indeed.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadPoolManager.java:
```diff
@@ -217,6 +260,107 @@ protected void afterExecute(Runnable runnable, Throwable throwable) {
+
+        @Override
+        public boolean remove(Runnable task) {
+            Runnable scheduledTask = scheduled.get(task);
```
Can't we use <code>Runnable scheduledTask = scheduled.remove(task);</code> here, as it will also return the object if there is one and we cannot loose / forget to remove it if there is an exception or something else?

kgoderis replied:
I thought It would be safer to first remove the task from the internal tracking collections, and then finally remove it at the end

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadPoolManager.java:
```diff
@@ -217,6 +260,107 @@ protected void afterExecute(Runnable runnable, Throwable throwable) {
+            if (futures.get(scheduledTask) != null && futures.get(scheduledTask).cancel(true)) {
+                running.remove(scheduledTask);
+            }
+            scheduled.remove(task);
```
Is there no need to remove scheduledTask of task from "futures"?

kgoderis replied:
yes indeed

maggu2810 commented:
Why do you add a "-" between the parameter name and the description in JavaDoc?
http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#@param

I see it in multiple files but don't know if there is any reason to do this.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,278 @@
+     */
+    public AbstractExpression(String expression, String delimiters, Date startDate, TimeZone timeZone,
+            int minimumCandidates) throws ParseException {
+        super();
```
Any reason to add this call as it is already done implicit?

kgoderis replied:
No - must be a leftover of something ;-)

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,278 @@
+
+        validateExpression();
+
+        if (startDate == null) {
```
see above

maggu2810 replied:
Ah, do you like that a sub-class could set startDate to null to ensure every time the current time is used?

kgoderis replied:
indeed

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,278 @@
+    abstract protected void validateExpression() throws IllegalArgumentException;
+
+    protected void applyExpressionParts() {
+        Collections.sort(getExpressionParts());
```
Above this is done in front of the validate call, so if it has been sorted before the validation, couldn't we use the expression part as it is instead of calling sort again...

kgoderis replied:
applyExpressionParts() could be called from elsewhere, one has to make sure it is sorted. I would then rather remove it from the validate call instead

maggu2810 commented:
Have to stop now (sorry) there are people waiting for me ;)
Will continue the review the next days (sorry again, but a lot of stuff to do).

The code I have seen looks good so far, just added some comments.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,278 @@
+
+        applyExpressionParts();
+
+        synchronized (this) {
```
You are using a synchronized block here, all fine.
In the applyExpressionsPart function you iterate over the expression parts without using some sync-mechanism that setExpressionsPart could be called.
As this is the only synchronized block are you sure that using this will make the objects of this classes thread safe?

kgoderis replied:
ok - maybe beter put a synchronzied() in setExpressionParts

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,278 @@
+
+    abstract protected E parseToken(String token, int position) throws ParseException;
+
+    abstract protected void populateWithSeeds();
```
Could you add a JavaDoc to this function as it has to be implemented by the sub class correctly.

kgoderis replied:
ok

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,278 @@
+        return getCandidates().get(getCandidates().size() - 1);
+    }
+
+    abstract protected E parseToken(String token, int position) throws ParseException;
```
Could you add a JavaDoc to this function as it has to be implemented by the sub class correctly.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java:
```diff
@@ -0,0 +1,278 @@
+            for (Date aDate : getCandidates()) {
+                logger.trace("Candidate {} is {}", getCandidates().indexOf(aDate), aDate);
+            }
+            prune();
```
Isn't it enough to call prune after the for-each loop or do we really need to call it every time?
If I understand the code correctly it should be enough to call it only once.
Also the logging loop should be enough to call it after the loop.

kgoderis replied:
It is debatable, as some Expressions can generate a LOT of candidates. Just imagine one for every second in the Calendar. I just added it in so that the actual subclasses can decide how they manage the avalanche of candidate dates.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpressionPart.java:
```diff
@@ -0,0 +1,180 @@
+
+    /**
+     * A <code>BoundedIntegerSet</code> is a set of Integers that is bound by a minimum and maximum value, that does
+     * allow or not allow negaive values, and that is either starts from index 0 or 1
```
typo: negaive => negative

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpressionPart.java:
```diff
@@ -0,0 +1,180 @@
+        private void validate(final Integer integer) {
+            if (negative) {
+                final int abs = Math.abs(integer);
+                if (abs < min || abs > max) {
```
I believe I do not understand this class at a first look.
Reading this lines of code I assume that min and max must be >= 0.
If min=5 and max=7 and negative is true, then -6 is an allowed integer value?

kgoderis replied:
yes indeed

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpressionPart.java:
```diff
@@ -0,0 +1,180 @@
+            if (start == null) {
+                startAt = this.min;
+            } else {
+                startAt = start;
```
You test end > max above but do not check for (start != null && start < min). Is this be done by intention?

kgoderis replied:
No it was not intended and should be corrected

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpressionPart.java:
```diff
@@ -0,0 +1,180 @@
+         * @param increment - increment to advance
+         * @return true if the addition of integers was successful
+         */
+        public boolean add(final Integer start, final Integer end, final Integer increment) {
```
To understand the logic completely, I need an example or more details about this class.

kgoderis replied:
Well, the general idea is that you parse a part of an expression, which should finally result in a series of dates that are included in the expression. So for example, for a day in the month, you would have a bound integer set with min = 1, max = 31. and thus an expression expressing something relative to a day in the month, could result in a set with the integers 5,6,7 to indicate that the 5th, 6th and the 7th day of the month are relevant for generating candidate dates. For months, you would accept integer values 1 to 12. and so forth. Some expressions work with relative time indications, e.g. -3 to indicate the third last of the given series (hence, negative values in the boundintegerset)

maggu2810 commented:
The reviewed all except the three expression implementations:
* CronExpression
* DateExpression
* RecurrenceExpression

@kgoderis Perhaps we can discuss my comments (and the BoundedIntegerSet before I start with further reading). Okay?

I tried to do the review thoroughly, so perhaps some pedantic comments. ;)

kgoderis commented:
@maggu2810 yep. I only have my iPhone nearby right now. Boundintegerset is used in the 2 of the 3 classes you have not reviewed. When you read those, it's usage will become clear.

kgoderis commented:
> I also noticed that you will have to take out the expression related code from the ThreadPoolManager.
> I'd suggest that I actually first port the new ThreadPoolManager implementation from jUPnP over and we can then see what you might need to extend for the expression support - my hope would be that you actually do not have to touch this class itself at all.

@kaikreuzer I presume it is https://github.com/jupnp/jupnp/pull/47  that you are referring to? the code I introduce is a ExpressionThreadPoolExecutor that extends ScheduledThreadPoolExecutor, whereas the jupnp code involves a QueueingThreadPoolExecutor that extends ThreadPoolExecutor. The base classes are quite different, unless you want to introduce the thread management part to the scheduled version of the base class as well ? If not, then it is easy. I only have to touch the class to support consistency in they way the static methods are named and used throughout the code base.....

kaikreuzer commented:
> whereas the jupnp code involves a QueueingThreadPoolExecutor that extends ThreadPoolExecutor.

You are absolutely right - I noticed the same when I tried porting the code on Friday.

> unless you want to introduce the thread management part to the scheduled version of the base class as well ? 

That was actually my intension, since we mainly have scheduled pools in ESH - and with the new thread pool implementation I would like to avoid problems like [this one](https://community.openhab.org/t/problem-with-thingmanager-and-threadpool/7365). But I noticed that it isn't straight forward to port the jUPnP code for this. So ok for me to use your changes for the scheduled pools now and we can then see how we can improve the queueing for them as well.

kgoderis commented:
@maggu2810 how do we proceed?

maggu2810 commented:
@kgoderis Sorry about the delay. Will read the remaining files today or tomorrow (will try to find the time). Is there anything to discuss about the previous comments?
@kaikreuzer Are you fine so it could be merged after @kgoderis and me passed the review or have you some comments?

kgoderis commented:
@maggu2810 I am ok with your comments, but did not yet process them

kaikreuzer commented:
> Are you fine so it could be merged after @kgoderis and me passed the review or have you some comments?

I so far left the review to you and @sjka - but I will have a rough look over it as well, once your comments are fully incorporated.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java:
```diff
@@ -0,0 +1,177 @@
+                        DateFormat df3 = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXXX");
+                        df3.setTimeZone(getTimeZone());
+                        theDate = df3.parse(getExpression());
+                    } catch (Exception g) {
```
Instead of the encapsulated expressions, I would prefer to use another function.
What do you think about e.g. this one:

```java
        /**
         * Try to parse using a set of valid formats.
         *
         * @param formats the date format strings in the order they should be used
         * @return true if parsing succeeded by a format, otherwise false
         */
        private boolean parseFormats(final String[] formats) {
            for (final String format : formats) {
                try {
                    final DateFormat df = new SimpleDateFormat(format);
                    df.setTimeZone(getTimeZone());
                    theDate = df.parse(getExpression());
                    return true;
                } catch (final Exception e) {
                    // Try next one...
                }
            }
            return false;
        }

        @Override
        public void parse() throws ParseException {
            // Unfortunately, the time zone formats available to SimpleDateFormat (Java 6 and earlier) are not ISO 8601
            // compliant. SimpleDateFormat understands time zone strings like "GMT+01:00" or "+0100", the latter
            // according to RFC # 822.
            //
            // Even if Java 7 added support for time zone descriptors according to ISO 8601, SimpleDateFormat is still
            // not able to properly parse a complete date string, as it has no support for optional parts.
            //
            // Reformatting the input string using regexp is certainly one possibility, but the replacement rules are
            // not as simple
            //
            // Some time zones are not full hours off UTC, so the string does not necessarily end with ":00".
            // ISO8601 allows only the number of hours to be included in the time zone, so "+01" is equivalent to
            // "+01:00". ISO8601 allows the usage of "Z" to indicate UTC instead of "+00:00".
            // The easier solution is to use the data type converter in JAXB, since JAXB must be able to parse
            // ISO8601 date string according to the XML Schema specification.

            // try {
            // Calendar cal = DatatypeConverter.parseDateTime(getExpression());
            // theDate = cal.getTime();
            // } catch (Exception e) {
            // throw new ParseException(getPart() + " is not an ISO8601 formatted date", 0);
            // }

            if (!parseFormats(
                    new String[] { "yyyy-MM-dd'T'HH:mm:ssX", "yyyy-MM-dd'T'HH:mm:ssXX", "yyyy-MM-dd'T'HH:mm:ssXXX" })) {
                throw new ParseException(getPart() + " is not an ISO8601 formatted date", 0);
            }
        }
```

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/CronExpression.java:
```diff
@@ -0,0 +1,1350 @@
+
+        public SecondsExpressionPart(String s) throws ParseException {
+            super(s);
+            parse();
```
A constructor should not call non-final functions.

maggu2810 commented:
I have only a short look over the files "CronExpression" and "RecurrenceExpression". The logic is / should be tested using unit tests. The unit tests will identify failures, etc.
So, I think you can start to integrate all comments.
After that perhaps @sjka would like to have another look at the code base.
At least @kaikreuzer could have a final look at the PR.

kgoderis commented:
@kaikreuzer @sjka @maggu2810 I have processed all comments that we agreed upon

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.scheduler/src/main/java/org/eclipse/smarthome/core/scheduler/common/ThreadPoolManager.java:
```diff
@@ -217,6 +260,108 @@ protected void afterExecute(Runnable runnable, Throwable throwable) {
+                throw new NullPointerException();
+            }
+
+            this.setCorePoolSize(this.getCorePoolSize() + 1);
```
This looks fundamentally flawed to me - when handing over a job for scheduling, you have to increase the pool size (btw, is it ever decreased again?) and you block the thread by a Thread.sleep until the time for execution is reached?
This is imho not at all the idea of a scheduler service. There should be a single thread responsible for tracking when which job needs to be executed and it must not make use of the thread pool itself for this.

Compare with ScheduledThreadPoolExecutor, which uses a delayedExecute() method, which puts the tasks into the internal DelayedWorkQueue for distributing it to the thread pool, when it is time to do so.

kgoderis replied:
You are too quick here I think. Expressions  can have multiple execution moments, so, in order to keep it simple there is one thread managing each expression, e.g. the scheduleTask.Putting the thread to sleep will not hurt the system either I think. IMHO things will get complicated if you want to put that in a single thread. Personally, I would refrain from interfering with the inner bodyworks of ScheduledThreadPoolExecutor (delayedExecute is private). If you put it in 1 thread, that thread would have to run like crazy in order to constantly evaluate the expression execution moments, no? In the best case you would have to evaluate all getTimeAfter(now) of all Expressions, and put it to sleep to the smallest time that comes up; and then you have to manage that thread every time a new expression is schedule()ed. 

On the pool size you might be right, but I don't remember if ScheduledThreadPoolExecutor automatically reduces that after a while.

kaikreuzer replied:
>  Expressions can have multiple execution moments

The same is true for the scheduleAtFixedRate jobs in the ScheduledThreadPoolExecutor.

> one thread managing each expression

This is exactly the no-go that I mean. This completely defeats the idea of a thread pool.

> Putting the thread to sleep will not hurt the system either I think

It does. Every single thread is expensive and the thread pools are there to reduce their number.

> would have to run like crazy in order to constantly evaluate the expression execution moments, no? 

No. Because you could only have to check your queue (which can be sorted by next execution time), if there is a job at the top of the queue that needs to be executed. If not, just sleep for until its startTime (or until a new job is added to the queue, which might be the top most job then).

kgoderis replied:
Q: What would be the concern of having the "managing thread" in the thread pool itself?

kaikreuzer replied:
No general concern, but it does not really make much sense, since this thread will be blocked eternally and thus be effectively not be part of the pool anyhow, right? I am not sure if I understand your argument for using a thread in the pool. I think having a dedicated thread makes it easier to give it access to the queue (the ones from the pool should not know about the queue) and also to give it an explicit name, so that it is easily identified in the debugger.

kgoderis commented:
I tried it with a few examples and it seems to work ;-) A few questions remain: 1. what if we have a large volume of expressions to evaluate. 2. what if we want to reuse the same task with different expressions, knowing that the "task" is the pivotal/unique object that makes the underlying scheduler tick, e.g. should we foresee the possibility to have the same task object being scheduled with different expressions or not?

kaikreuzer commented:
> 1. what if we have a large volume of expressions to evaluate.

We should probably keep a sorted Map of tasks to not having to iterate through the whole list again and again. Not sure how the normal scheduler handles this, maybe it is worth to use it as a source for ideas.

> 2. what if we want to reuse the same task with different expressions

What problem is there, if a user passes the same instance with two different expressions to the scheduler? Does Quartz support this?

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/scheduler/ThreadPoolManager.java:
```diff
@@ -5,16 +5,25 @@
  * http://www.eclipse.org/legal/epl-v10.html
  */
-package org.eclipse.smarthome.core.common;
+package org.eclipse.smarthome.core.scheduler;
```
This class should imho stay where it was, so that we keep backward compatibility.
So we could either 
a) have a dependency from core.common to core.scheduler by this
b) have getExpressionScheduledPool() return a ThreadPoolExecutor that needs to be casted by the caller
c) Introduce a core.scheduler.ExpressionThreadPoolManager class that extends the normal one.

No solution looks ideal to me, so discussion is opened :-)

kgoderis replied:
I am open for anything basically. I will only have 1 problem in the coming week(s) and that is a lack of time, so I am open for you/someone else to take over the PR to get it finished.

kaikreuzer replied:
I think option b is probably the quickest intermediate solution then. So could you move the class back and squash all your commits? I can then try to take over further processing (probably soon creating a CQ as this might also take a while to be approved). Thanks!

kgoderis commented:
Can we finish this PR as other work depends on it?

kaikreuzer commented:
It looks as if I didn't express myself right in https://github.com/eclipse/smarthome/pull/959#discussion_r55547167 and you have now moved ALL classes from scheduler to common. In my suggestion, I was ONLY talking about the class ThreadPoolExecutor that needs to stay in common in order not to break any backward compatibility. All the rest is supposed to be in scheduler as already [discussed before](https://github.com/eclipse/smarthome/pull/959#issuecomment-180738971).

kaikreuzer commented:
@maggu2810 Are you fine with this PR in general? I would then create the CQ once @kgoderis has moved the classes back to the scheduler package.

maggu2810 commented:
If I look at the "Files changed" view there is still commented stuff, that needs to be changed (IMHO).

kgoderis commented:
@kaikreuzer heu? I reverted the commit and squashed the whole lot. Dunno understand. In order to avoid further confusion, I suggest you take over and put it they way you think is best.....

kaikreuzer commented:
> I reverted the commit

All I asked for was that "This class (ThreadPoolManager) should imho stay where it was".

> I suggest you take over and put it they way you think is best.....

Ok, I'll see when I find the time...

kgoderis commented:
@kaikreuzer cleaned up. Only the Quartz related questions are still to be discussed.

kaikreuzer commented:
I still do not see the core.scheduler package being back. Never mind, I will try to find the time to do it myself soon...

kgoderis commented:
I reverted it back to the original situation before all the back/fwd movements

kaikreuzer commented:
Ok, so I have created https://github.com/kgoderis/smarthome/pull/4, which you can merge and squash - once done, I will create the CQ.

kgoderis commented:
Opened a new clean PR

maggu2810 commented:
Superseded by #1239
